beater/middleware/middleware_test.go

```diff
@@ -33,22 +33,29 @@ func TestWithMiddleware(t *testing.T) {
 		c.Result.Keyword += s
 	}
 	h := func(c *request.Context) { addKeyword(c, "[h]") }
-	m1 := func(h request.Handler) request.Handler {
+	m1 := func(h request.Handler) (request.Handler, error) {
 		return func(c *request.Context) {
 			addKeyword(c, "[m1.1]")
 			h(c)
 			addKeyword(c, "[m1.2]")
-		}
+		}, nil
 	}
-	m2 := func(h request.Handler) request.Handler {
+	m2 := func(h request.Handler) (request.Handler, error) {
 		return func(c *request.Context) {
 			addKeyword(c, "[m2.1]")
 			h(c)
 			addKeyword(c, "[m2.2]")
-		}
+		}, nil
 	}
 	c, _ := beatertest.DefaultContextWithResponseRecorder()
-	Wrap(h, m1, m2)(c)
+	h, e := Wrap(h, m1, m2)
+	require.NoError(t, e)
+	h(c)
 	require.NotNil(t, c.Result.Keyword)
 	assert.Equal(t, "[m1.1][m2.1][h][m2.2][m1.2]", c.Result.Keyword)
 }
+
+func Apply(m Middleware, h request.Handler) request.Handler {
+	r, _ := m(h)
+	return r
```

Comment:
I don't see why a middleware can't return an error if it fails to initialize itself - as a matter of fact I think this makes the general handling more flexible. Even more so, I think the `request.Handler` should also return an error to be propagated downstream, instead of maintaining some state in the context - but didn't want to make more changes than strictly necessary.
(And this is just a test helper by the way)

Reply:
As mentioned https://github.com/elastic/apm-server/pull/2706#discussion_r324140270 I was misreading the level of where the error was returned. My comment was referring to that error and not meant as a general comment on a middleware not returning an error.